### PR TITLE
MULTIARCH-3743: Add support of PowerVS on hypershift-hosted workflow

### PR DESCRIPTION
**Pull request by dharaneeshvrd**

### CODE REVIEW

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-powervs-e2e

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-powervs-e2e

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-aws

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-powervs-e2e
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-aws

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-aws

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-powervs-e2e

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-powervs-e2e

**Comment by bear-redhat** [comment]
/pj-hearse pull-ci-redhat-appstudio-infra-deployments-main-appstudio-e2e-tests

**Comment by bear-redhat** [comment]
/pj-rehearse /pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-aws

**Comment by bear-redhat** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-hypershift-hostedcluster-aws

**Comment by bear-redhat** [comment]
/pj-rehearse pull-ci-redhat-appstudio-e2e-tests-main-redhat-appstudio-e2e

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh [hunk 1]
```diff
@@ -9,11 +9,21 @@ oc registry login --to=${SHARED_DIR}/pull-secret-build-farm.json
...
-if [[ ! -f "${AWS_GUEST_INFRA_CREDENTIALS_FILE}" ]]; then
-  echo "AWS credentials file ${AWS_GUEST_INFRA_CREDENTIALS_FILE} not found"
+if [[ "${PLATFORM}" == "aws" ]]; then
+  AWS_GUEST_INFRA_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred"
+  if [[ ! -f "${AWS_GUEST_INFRA_CREDENTIALS_FILE}" ]]; then
+    echo "AWS credentials file ${AWS_GUEST_INFRA_CREDENTIALS_FILE} not found"
+    exit 1
+  fi
+elif [[ "${PLATFORM}" == "powervs" ]]; then
+  export IBMCLOUD_CREDENTIALS="${CLUSTER_PROFILE_DIR}/credentials"
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:19]: - need to add a check here for this file before exporting it
- name of the credential file can be more specific to platform and also hide like aws `.powervs`

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml [hunk 2]
```diff
@@ -2,6 +2,9 @@ ref:
   as: hypershift-hostedcluster-create-hostedcluster
   cli: latest
   env:
+  - name: PLATFORM
+    default: "aws"
+    documentation: "The platform(AWS/PowerVS) of the HostedCluster. AWS by default"
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml:7]: lets document the exact the options we are using
```suggestion
    documentation: "The platform(aws/powervs) of the HostedCluster. aws by default"
```

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml [hunk 3]
```diff
@@ -38,6 +41,36 @@ ref:
...
+    documentation: "The system type of Power nodes."
+  - name: POWERVS_GUID
+    default: ""
+    documentation: "IBM Cloud PowerVS Service Instance ID.This flag is used to reuse an existing PowerVS Service Instance resource for cluster's infra"
+  - name: VPC
+    default: ""
+    documentation: "IBM Cloud VPC Name. This flag is used to reuse an existing VPC resource for cluster's infra"
+  - name: CLOUD_CONNECTION
+    default: ""
+    documentation: "Cloud Connection in given zone.This flag is used to reuse an existing Cloud Connection resource for cluster's infra"
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml:73]: it will be good to prefix all the environment variable with `POWERVS` to easily identify for the platform

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml [hunk 4]
```diff
@@ -59,11 +92,12 @@ ref:
...
+    name: registry-pull-credentials
+    namespace: test-credentials
   - mount_path: /var/run/hypershift-workload-credentials
     name: hypershift-workload-credentials
     namespace: test-credentials
   dependencies:
   - name: "release:latest"
     env: RELEASE_IMAGE_LATEST
-  - name: "stable:hypershift"
-    env: HYPERSHIFT_RELEASE_LATEST
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml:69]: why is this removed?

File: ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh [hunk 5]
```diff
@@ -32,27 +38,51 @@ else
 fi
 
 createdAt=`oc -n clusters get hostedclusters $CLUSTER_NAME -o jsonpath='{.metadata.annotations.created-at}'`
-if [ -z $createdAt ]; then
+if [ -z $createdAt ] && [ "${PLATFORM}" == "aws" ]; then
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh:41]: how this if block is only specific to `aws` platform?

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh [hunk 6]
```diff
@@ -46,28 +54,59 @@ if [[ "${COMPUTE_NODE_TYPE}" == "" ]]; then
...
+      --annotations hypershift.openshift.io/cleanup-cloud-resources="false" \
+      --additional-tags "prow.k8s.io/job=${JOB_NAME}" \
+      --additional-tags "prow.k8s.io/build-id=${BUILD_ID}" 
+    ;;
+  "powervs")
+    bin/hypershift create cluster powervs \
+      --name ${CLUSTER_NAME} \
+      --infra-id ${INFRA_ID} \
+      --node-pool-replicas ${HYPERSHIFT_NODE_COUNT} \
+      --base-domain ${HYPERSHIFT_BASE_DOMAIN} \
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:87]:
```suggestion
      --base-domain ${DOMAIN} \
```

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh [hunk 7]
```diff
@@ -46,28 +54,59 @@ if [[ "${COMPUTE_NODE_TYPE}" == "" ]]; then
...
+    /usr/bin/hypershift create cluster aws \
+      ${EXTRA_ARGS} \
+      --name ${CLUSTER_NAME} \
+      --infra-id ${INFRA_ID} \
+      --node-pool-replicas ${HYPERSHIFT_NODE_COUNT} \
+      --instance-type=${COMPUTE_NODE_TYPE} \
+      --base-domain ${DOMAIN} \
+      --region ${HYPERSHIFT_AWS_REGION} \
+      --control-plane-availability-policy ${HYPERSHIFT_CP_AVAILABILITY_POLICY} \
+      --infra-availability-policy ${HYPERSHIFT_INFRA_AVAILABILITY_POLICY} \
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:68]: do you wanna use these params for the powervs platform as well?

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh [hunk 8]
```diff
@@ -46,28 +54,59 @@ if [[ "${COMPUTE_NODE_TYPE}" == "" ]]; then
...
+      --pull-secret=/tmp/pull-secret.json \
+      --aws-creds=${AWS_GUEST_INFRA_CREDENTIALS_FILE} \
+      --release-image ${RELEASE_IMAGE} \
+      --control-plane-operator-image=${CONTROLPLANE_OPERATOR_IMAGE:-} \
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:72]: do you wanna use this param for the powervs platform as well?

**Review by mkumatag — CHANGES_REQUESTED** [review]

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-powervs

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-powervs

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-powervs

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-powervs

**Comment by dharaneeshvrd** [comment]
/pj-rehearse periodic-ci-openshift-hypershift-release-4.14-periodics-e2e-powervs

**Comment by dharaneeshvrd** [comment]
/pj-rehearse pull-ci-redhat-appstudio-e2e-tests-main-redhat-appstudio-e2e

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh (hunk 7 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:72]: yes, used.

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh (hunk 8 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh:76]: yes, used.

File: ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh (hunk 5 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh:45]: Its blocked to debug the failed cluster by dptp team, since we don't have permission to do much debugging, its better to unblock the destroy for powervs platform.

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh (hunk 1 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh]: Followed aws approach, added a check and renamed the cress file.

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml (hunk 2 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml]: Done

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml (hunk 3 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml]: Done

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml (hunk 4 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml]: Reverted back

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh (hunk 6 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-commands.sh]: Done

**Comment by dharaneeshvrd** [comment]
@mkumatag Addressed your comments, ptal!

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml [hunk 9]
```diff
@@ -59,11 +92,14 @@ ref:
...
+    namespace: test-credentials
   - mount_path: /var/run/hypershift-workload-credentials
     name: hypershift-workload-credentials
     namespace: test-credentials
   dependencies:
   - name: "release:latest"
     env: RELEASE_IMAGE_LATEST
   - name: "stable:hypershift"
-    env: HYPERSHIFT_RELEASE_LATEST
+    env: HYPERSHIFT_RELEASE_LATEST
```
Comment (mkumatag) [inline ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml:105]: Add a newline in the end

**Review by mkumatag — CHANGES_REQUESTED** [review]
one last minor nit; otherwise lgtm

File: ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml (hunk 9 above)
Comment (dharaneeshvrd) [reply ci-operator/step-registry/hypershift/hostedcluster/create/hostedcluster/hypershift-hostedcluster-create-hostedcluster-ref.yaml]: Fixed

File: ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh (hunk 5 above)
Comment (bear-redhat) [reply ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh:45]: @dharaneeshvrd  just want to make sure you aware, for some broken clusters, the destroy command cannot properly remove them. Thus, we have another periodic job to remove those broken clusters every 24 hours. We have other periodics to remove AWS resources created more than certain numbers of hours too, as a backup.

**Review by bear-redhat — APPROVED** [review]

File: ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh (hunk 5 above)
Comment (mkumatag) [reply ci-operator/step-registry/hypershift/hostedcluster/destroy/hostedcluster/hypershift-hostedcluster-destroy-hostedcluster-commands.sh:45]: @dharaneeshvrd lets add that flow as well as a follow up PR
